sessions/2018/november/session_16.md

```diff
@@ -0,0 +1,89 @@
+# SOFTWARE ENGINEERING IN JAVA
+
+## Session 16 (06/11/2018)
+
+- Optimizations to Homework code
+
+### Notes:
+
+This session was a reviewing of the homework code, specially the part dealing with fibonacci series:
+
+```java
+public List<Integer> getFibonacci(Integer number) {
+     List<Integer> fibonacci = new ArrayList<>();
+    fibonacci.add(1);
+    fibonacci.add(1);
+     for (int i = 2; i < number; i++) {
+        fibonacci.add(i, fibonacci.get(i - 1) + fibonacci.get(i - 2));
+    }
+     return fibonacci;
+ }
+```
+
+For that code, linear time complexity `O(n)` is used, that means that the running time will increase according to the input number.
+
+The code was upgrade to the following:
+
+```java
+@Service
+public class Calculator {
+
+    List<BigInteger> fibonacci = new ArrayList<>();
+
+    public List<BigInteger> getFibonacci(Integer number) {
+
+        if (fibonacci.size() >= number){
+            return fibonacci.subList(0,number);
+
+        } else {
+            if (fibonacci.size() == 0) {
+                fibonacci.add(BigInteger.ONE);
+                fibonacci.add(BigInteger.ONE);
+
+                for (int i = 2; i < number; i++) {
+                    fibonacci.add(i, fibonacci.get(i - 1).add(fibonacci.get(i - 2)));
+                }
+            }
+
+            for (int i = fibonacci.size(); i < number; i++) {
+                fibonacci.add(i, fibonacci.get(i - 1).add(fibonacci.get(i - 2)));
+            }
+
+        }
+
+        return fibonacci;
+
+    }
+```
+When the list in empty the time complexity is `O(n)`, but in order to improve the time complexity for subsequent queries, we're using a cache notion. So instead to begin to calculate from scratch the given number, we're using our previous fibonacci calculated list to speed up the answer. If the new given number is less or equal than the cached list size, we return inmediately the corresponding sublist and our time complexity becomes constant `O(1)`. By other side, if the number is major than the cached list, we calculate just the missing values, and even if the time complexity continues being `O(n)', the answer will be quicker because less steps are neccesary to do.
```

Comment:
Just some improvements on the English in this bit:

```suggestion
 When the list is empty the time complexity is `O(n)`, but in order to improve the time complexity for subsequent queries, we're using a cache. Instead of beginning the calculation from scratch the each time, we're using our previous fibonacci calculated list to speed up the answer. If the new given number is less or equal than the cached list size, we return immediately the corresponding sublist and our time complexity becomes constant `O(1)`. Conversely, if the number is larger than the size of the cached list, we calculate just the missing values, and even if the time complexity continues being `O(n)', the answer will be quicker because the intermediary steps are cached.
```